# 利用Keras探索声呐物体分类数据集

## 目录

- [1. 数据集介绍](#1.-数据集介绍)
- [2. 数据加载分析](#2.-数据加载分析)
- [3. 数据处理](#3.-数据处理)

# 1. 数据集介绍

本章使用声呐数据，包括声呐在不同物体的返回。数据有60个变量，代表不同角度的返回值。目标是将石头和金属筒（矿石）分开。

所有的数据都是连续的，从0到1；输出变量中M代表矿石，R代表石头，需要转换为1和0。数据集有208条数据

**数据下载地址**: http://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

# 2. 数据加载分析

In [52]:
import pandas as pd
import numpy as np

dataset = pd.read_csv('../data/sonar.csv', header=None)

In [53]:
dataset.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9  ...      51      52      53      54      55      56      57      58  \
0  0.2111 ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  0.0090   
1  0.2872 ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049  0.0052   
2  0.6194 ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164  0.0095   
3  0.1264 ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044  0.0040   
4  0.4459 ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048  0.0107   

       59  60  
0  0.0032   R  
1  0.0044   R  
2  0.0078   R  
3  0.0117   R  
4  0.0094   R  

[5 rows x 61 columns]

In [54]:
dataset.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9      ...              50  \
count  208.000000  208.000000  208.000000  208.000000     ...      208.000000   
mean     0.121747    0.134799    0.178003    0.208259     ...        0.016069   
std      0.061788    0.085152    0.118387    0.134416     ...        0.012008   
min      0.003300    0.005500    0.007500    0.011300     ...        0.000000   
25%      0.080900    0.080425    0.097025    0.111275     ...        0.008425   
50%      0.106950    0.112100    0.152250    0.182400     ...        0.013900   
75%      0.154000    0.169600    0.233425    0.268700     ...        0.020825   
max      0.372900    0.459000    0.682800    0.710600     ...        0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [55]:
dataset.describe().transpose()

count      mean       std     min       25%      50%       75%     max
0   208.0  0.029164  0.022991  0.0015  0.013350  0.02280  0.035550  0.1371
1   208.0  0.038437  0.032960  0.0006  0.016450  0.03080  0.047950  0.2339
2   208.0  0.043832  0.038428  0.0015  0.018950  0.03430  0.057950  0.3059
3   208.0  0.053892  0.046528  0.0058  0.024375  0.04405  0.064500  0.4264
4   208.0  0.075202  0.055552  0.0067  0.038050  0.06250  0.100275  0.4010
5   208.0  0.104570  0.059105  0.0102  0.067025  0.09215  0.134125  0.3823
6   208.0  0.121747  0.061788  0.0033  0.080900  0.10695  0.154000  0.3729
7   208.0  0.134799  0.085152  0.0055  0.080425  0.11210  0.169600  0.4590
8   208.0  0.178003  0.118387  0.0075  0.097025  0.15225  0.233425  0.6828
9   208.0  0.208259  0.134416  0.0113  0.111275  0.18240  0.268700  0.7106
10  208.0  0.236013  0.132705  0.0289  0.129250  0.22480  0.301650  0.7342
11  208.0  0.250221  0.140072  0.0236  0.133475  0.24905  0.331250  0.7060
12  208.0  0.273305  0.140962  0.0184  0.166125  0.26395  0.351250  0.7131
13  208.0  0.296568  0.164474  0.0273  0.175175  0.28110  0.386175  0.9970
14  208.0  0.320201  0.205427  0.0031  0.164625  0.28170  0.452925  1.0000
15  208.0  0.378487  0.232650  0.0162  0.196300  0.30470  0.535725  0.9988
16  208.0  0.415983  0.263677  0.0349  0.205850  0.30840  0.659425  1.0000
17  208.0  0.452318  0.261529  0.0375  0.242075  0.36830  0.679050  1.0000
18  208.0  0.504812  0.257988  0.0494  0.299075  0.43495  0.731400  1.0000
19  208.0  0.563047  0.262653  0.0656  0.350625  0.54250  0.809325  1.0000
20  208.0  0.609060  0.257818  0.0512  0.399725  0.61770  0.816975  1.0000
21  208.0  0.624275  0.255883  0.0219  0.406925  0.66490  0.831975  1.0000
22  208.0  0.646975  0.250175  0.0563  0.450225  0.69970  0.848575  1.0000
23  208.0  0.672654  0.239116  0.0239  0.540725  0.69850  0.872175  1.0000
24  208.0  0.675424  0.244926  0.0240  0.525800  0.72110  0.873725  1.0000
25  208.0  0.699866  0.237228  0.0921  0.544175  0.75450  0.893800  1.0000
26  208.0  0.702155  0.245657  0.0481  0.531900  0.74560  0.917100  1.0000
27  208.0  0.694024  0.237189  0.0284  0.534775  0.73190  0.900275  1.0000
28  208.0  0.642074  0.240250  0.0144  0.463700  0.68080  0.852125  1.0000
29  208.0  0.580928  0.220749  0.0613  0.411400  0.60715  0.735175  1.0000
30  208.0  0.504475  0.213992  0.0482  0.345550  0.49035  0.641950  0.9657
31  208.0  0.439040  0.213237  0.0404  0.281400  0.42960  0.580300  0.9306
32  208.0  0.417220  0.206513  0.0477  0.257875  0.39120  0.556125  1.0000
33  208.0  0.403233  0.231242  0.0212  0.217575  0.35105  0.596125  0.9647
34  208.0  0.392571  0.259132  0.0223  0.179375  0.31275  0.593350  1.0000
35  208.0  0.384848  0.264121  0.0080  0.154350  0.32115  0.556525  1.0000
36  208.0  0.363807  0.239912  0.0351  0.160100  0.30630  0.518900  0.9497
37  208.0  0.339657  0.212973  0.0383  0.174275  0.31270  0.440550  1.0000
38  208.0  0.325800  0.199075  0.0371  0.173975  0.28350  0.434900  0.9857
39  208.0  0.311207  0.178662  0.0117  0.186450  0.27805  0.424350  0.9297
40  208.0  0.289252  0.171111  0.0360  0.163100  0.25950  0.387525  0.8995
41  208.0  0.278293  0.168728  0.0056  0.158900  0.24510  0.384250  0.8246
42  208.0  0.246542  0.138993  0.0000  0.155200  0.22255  0.324525  0.7733
43  208.0  0.214075  0.133291  0.0000  0.126875  0.17770  0.271750  0.7762
44  208.0  0.197232  0.151628  0.0000  0.094475  0.14800  0.231550  0.7034
45  208.0  0.160631  0.133938  0.0000  0.068550  0.12135  0.200375  0.7292
46  208.0  0.122453  0.086953  0.0000  0.064250  0.10165  0.154425  0.5522
47  208.0  0.091424  0.062417  0.0000  0.045125  0.07810  0.120100  0.3339
48  208.0  0.051929  0.035954  0.0000  0.026350  0.04470  0.068525  0.1981
49  208.0  0.020424  0.013665  0.0000  0.011550  0.01790  0.025275  0.0825
50  208.0  0.016069  0.012008  0.0000  0.008425  0.01390  0.020825  0.1004
51  208.0  0.013420  0.009634  0.0008  0.007275  0.01140  0.016725  0.0709
52  208.0  0.010709  0.007060

In [56]:
dataset.iloc[:,60].unique()

array(['R', 'M'], dtype=object)

In [57]:
print dataset.groupby(60).size()

60
M    111
R     97
dtype: int64


"sonar.csv"中，只有两种类别
- R: 石头
- M: 矿石

# 3. 构建一个简单的网络

In [96]:
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.cross_validation import cross_val_score, StratifiedKFold
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.pipeline import Pipeline


In [94]:
seed = 60
np.random.seed(seed)

In [80]:
X = dataset.iloc[:,0:60].values
y = dataset.iloc[:,60].values


[[0.02   0.0371 0.0428 ... 0.0084 0.009  0.0032]
 [0.0453 0.0523 0.0843 ... 0.0049 0.0052 0.0044]
 [0.0262 0.0582 0.1099 ... 0.0164 0.0095 0.0078]
 ...
 [0.0522 0.0437 0.018  ... 0.0138 0.0077 0.0031]
 [0.0303 0.0353 0.049  ... 0.0079 0.0036 0.0048]
 [0.026  0.0363 0.0136 ... 0.0036 0.0061 0.0115]]


将 y 进行 **ONE-HOT** encoder

In [70]:
encoder = LabelEncoder()
Y = encoder.fit_transform(y)

#Y = pd.get_dummies(y1).values

In [101]:
def baseline_model():
    model = Sequential()
    model.add(Dense(100, input_dim=60, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [102]:
estimator = KerasClassifier(build_fn=baseline_model, nb_epoch=150, batch_size=5, verbose=0)
kfold = StratifiedKFold(y=Y, n_folds=10, shuffle=True, random_state=seed)
scores = cross_val_score(estimator, X, Y, cv=kfold)
print 'Results %.2f %% (%.2f %%)' % (scores.mean()*100, scores.std()*100)

Results 64.98 % (8.28 %)


accuracy 为: 64.98%, 标准差为: 8.28%, 效果也不是那么的差，因为实现的代码确实很简单

# 4. 预处理数据增加性能 

预处理数据是个好习惯。神经网络喜欢输入类型的比例和分布一致，为了达到这点可以使用正则化，让数据的平均值是0，标准差是1，这样可以保留数据的分布情况。

scikit-learn的StandardScaler可以做到这点。不应该在整个数据集上直接应用正则化：应该只在测试数据上交叉验证时进行正则化处理，使正则化成为交叉验证的一环，让模型没有新数据的先验知识，防止模型发散。

scikit-learn的Pipeline可以直接做到这些。我们先定义一个 StandardScaler，然后进行验证：

In [105]:
estimaters = []
estimaters.append(('standardize', StandardScaler()))
estimaters.append(('mlp', KerasClassifier(build_fn=baseline_model, nb_epoch=150, batch_size=5, verbose=0)))
pipeline = Pipeline(estimaters)

kfold = StratifiedKFold(y=Y, n_folds=10, shuffle=True, random_state=seed)
scores = cross_val_score(pipeline, X, Y, cv=kfold)
print 'Standardized %.2f %% (%.2f %%)' % (scores.mean()*100, scores.std()*100)

Standardized 75.01 % (6.97 %)


增加了 Standardize 的确提高了 accuracy, 降低了 标准差

# 5. 调整模型的拓扑和神经元

神经网络有很多参数，例如初始化权重、激活函数、优化算法等等。我们一直没有说到调整网络的拓扑结构：扩大或缩小网络。我们试验一下：


## 4.1 缩小网络

有可能数据中有冗余：原始数据是不同角度的信号，有可能其中某些角度有相关性。我们把第一层隐层缩小一些，强行提取特征试试。

我们把之前的模型隐层的100个神经元减半到50个，这样神经网络需要挑选最重要的信息。之前的正则化有效果,也一并在这里做一下.

In [109]:
def baseline_model():
    model = Sequential()
    model.add(Dense(50, input_dim=60, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

np.random.seed(seed)
estimaters = []
estimaters.append(('standardize', StandardScaler()))
estimaters.append(('mlp', KerasClassifier(build_fn=baseline_model, nb_epoch=150, batch_size=5, verbose=0)))
pipeline = Pipeline(estimaters)

kfold = StratifiedKFold(y=Y, n_folds=10, shuffle=True, random_state=seed)
scores = cross_val_score(pipeline, X, Y, cv=kfold)
print 'Standardized %.2f %% (%.2f %%)' % (scores.mean()*100, scores.std()*100)

Standardized 74.15 % (11.23 %)


## 4.2 扩大网络
扩大网络后，神经网络更有可能提取关键特征，以非线性方式组合。

In [116]:
def baseline_model():
    model = Sequential()
    model.add(Dense(120, input_dim=60, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(30, activation='relu'))
    model.add(Dense(15, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

np.random.seed(seed)
# estimaters = []
# estimaters.append(('standardize', StandardScaler()))
# estimaters.append(('mlp', KerasClassifier(build_fn=baseline_model, nb_epoch=150, batch_size=20, verbose=0)))
# pipeline = Pipeline(estimaters)

estimator = KerasClassifier(build_fn=baseline_model, nb_epoch=50, batch_size=10, verbose=0)
kfold = StratifiedKFold(y=Y, n_folds=5, shuffle=True, random_state=seed)
scores = cross_val_score(estimator, X, Y, cv=kfold)
print 'Standardized %.2f %% (%.2f %%)' % (scores.mean()*100, scores.std()*100)

Standardized 61.50 % (8.02 %)
